src/tools/scripts/run_and_test_website.sh

```diff
@@ -48,19 +48,17 @@ fi
 
 if [ "$(pgrep -f 'python main.py')" ]; then
   echo "Killing existing server to run a fresh version"
-  pkill -9 python main.py
+  pkill -9 -f "python main.py"
+fi
+
+if [ "$(pgrep -f 'node ./tools/generate/chapter_watcher')" ]; then
+  echo "Killing existing watcher to run a fresh version"
+  pkill -9 -f "node ./tools/generate/chapter_watcher"
 fi
 
 echo "Installing and testing python environment"
 python -m pip install --upgrade pip
```

Comment:
This line is causing `pip` to be installed again every time a container is run, which means the version in the Docker image is not the latest.

```
$ docker container run --rm -it -v /app/node_modules -v "$PWD":/app -p 8080:8080 webalmanac

> almanac.httparchive.org@0.0.1 start /app
> run-script-os


> almanac.httparchive.org@0.0.1 start:darwin:linux /app
> ./tools/scripts/run_and_test_website.sh -d

Installing and testing python environment
Collecting pip
  Downloading pip-20.2.4-py2.py3-none-any.whl (1.5 MB)
     |████████████████████████████████| 1.5 MB 3.6 MB/s 
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.1
    Uninstalling pip-20.1:
      Successfully uninstalled pip-20.1
Successfully installed pip-20.2.4
```

Reply:
If an image was built a while ago `pip` could be out of date. By doing this command it will try to install the latest and if that was what the container was built with, then no download and no issue:

```
Installing and testing python environment
Requirement already up-to-date: pip in /usr/local/lib/python3.9/site-packages (20.2.4)
Requirement already satisfied: flask==1.1.2 in /usr/local/lib/python3.9/site-packages (from -r requirements.txt (line 1)) (1.1.2)
Requirement already satisfied: flask-talisman==0.7.0 in /usr/local/lib/python3.9/site-packages (from -r requirements.txt (line 2)) (0.7.0)
...
```

And if it wasn't then time to update your container to avoid this slower way of starting. 😁

Either way it will work - just will be faster if container was built with latest pip.

Does that make sense? Can you see any issue with that?

Reply:
I would prefer if we do not install things in the image at runtime that are needed for the lifecycle of a contain, but make sure the image has them already. This way the image is more self-contained, portable, and reproducible. If a dependency or package is old, it needs to be escalated in the image building time and a fresh image should be built so that at run time we have everything we need.

Reply:
Going to Resolve this and move to the general conversation as two different threads on basically the same topic.